userd/settings.go

```diff
@@ -0,0 +1,125 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package userd
+
+import (
+	"fmt"
+	"os/exec"
+	"strings"
+
+	"github.com/godbus/dbus"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/osutil"
+)
+
+const settingsIntrospectionXML = `
+<interface name="org.freedesktop.DBus.Peer">
+	<method name='Ping'>
+	</method>
+	<method name='GetMachineId'>
+               <arg type='s' name='machine_uuid' direction='out'/>
+	</method>
+</interface>
+<interface name='io.snapcraft.Settings'>
+	<method name='Check'>
+		<arg type='s' name='setting' direction='in'/>
+		<arg type='s' dbusInterfacename='check' direction='in'/>
+                <arg type='s' name='result' direction='out'/>
+	</method>
+	<method name='Get'>
+		<arg type='s' name='setting' direction='in'/>
+                <arg type='s' name='result' direction='out'/>
+	</method>
+	<method name='Set'>
+		<arg type='s' name='setting' direction='in'/>
+		<arg type='s' name='value' direction='in'/>
+	</method>
+</interface>`
+
+// Settings implements the 'io.snapcraft.Settings' DBus interface.
+type Settings struct{}
+
+// Name returns the name of the interface this object implements
+func (s *Settings) Name() string {
+	return "io.snapcraft.Settings"
+}
+
+// BasePath returns the base path of the object
+func (s *Settings) BasePath() dbus.ObjectPath {
+	return "/io/snapcraft/Settings"
+}
+
+// IntrospectionData gives the XML formatted introspection description
+// of the DBus service.
+func (s *Settings) IntrospectionData() string {
+	return settingsIntrospectionXML
+}
+
+// Check implements the 'Check' method of the 'com.canonical.Settings'
+// DBus interface.
+//
+// Example usage: dbus-send --session --dest=io.snapcraft.Settings --type=method_call --print-reply /io/snapcraft/Settings io.snapcraft.Settings.Check string:'default-web-browser' string:'firefox.desktop'
+func (s *Settings) Check(setting, check string) (string, *dbus.Error) {
+	cmd := exec.Command("xdg-settings", "check", setting, check)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", dbus.MakeFailedError(fmt.Errorf("cannot check setting %s: %s", setting, osutil.OutputErr(output, err)))
+	}
+
+	return strings.TrimSpace(string(output)), nil
```

Comment:
The whole check/get/set paradigm doesn't map directly to snaps wrt info-disclosure. I think we can handle this though since userd can mediate.
* if 'Check()', then see if the value of 'check' matches one of the desktop files of the snap. If not, return error, if so, send it along to the system's xdg-settings and return the result
* likewise if Set(), verify if 'new' belongs to the snap. If not, return an error, if so, send it along to the system's xdg-settings and return the result
* Get() is slightly harder but doable. Run the get request on the system's xdg-settings and see if the result belongs to the snap, if so, return the result, if not, mock up something (eg, `NOT_THIS_SNAP.snap.desktop` (or similar-- I chose uppercase and an extra '_' because that is an invalid snap name/command combination for a desktop file)).

Reply:
The question then becomes how to answer what snap the connecting process is coming from. You can either do this with libapparmor or see which snap cgroup freezer the pid is in. The former is easier to implement and most correct, but the latter works cross-distro where apparmor isn't available.

Reply:
Silly question - but how terrible is the information disclosure we have here in Get/Check? Given that we only allow to query for "default-web-browser" currently and given that the answer is a string like "firefox.desktop" I wonder if that is an acceptable risk? As for Set() this is currently behind a dialog so no snap can sneak behind the users back. But I see that for Set() we may want to protect things a bit more. I can see a (weak) use case for a browser where e.g. the browser has "set $browser as default" and if the user sets it a option "revert to previous setting" where it does make sense to allow a snap to set something else than itself.

Reply:
The information leak is not terrible in this case, but avoidable. One day I'd like to be in a position where there or no information leaks, so I try to avoid introducing any new whenever feasible.